File: python/bracket-push/bracket_push.py
def is_paired(input_string):
    stack = list()
    brackets = {'{': '}', '[': ']', '(': ')' }

    for i in input_string:
        if i in brackets.keys():
            stack.append(i)
            
        elif i in brackets.values():
            if not stack:
                return False
            
            stack_top = stack.pop()

            if i != brackets[stack_top]:
                return False

    return not stack

File: cpp/bracket-push/bracket_push.cpp
#include "bracket_push.h"

char getOpeningBracket(char c) {
        if (c == ']') {
                return '[';
        }
        else if (c == '}') {
                return '{';
        }
        else if (c == ')') {
                return '(';
        }
        else {
                throw std::invalid_argument("Invalid input to getOpeningBracket"); 
        }
}

bool bracket_push::check(std::string input)
{
        std::stack<char> s; 

        for (char c : input) {
                if (c == '[' || c == '{' || c == '(') {
                        s.push(c); 
                }
                else if (c == ']' || c == '}' || c == ')') {

                        if (s.empty()) {
                                return false; 
                        }

                        char opening_bracket = getOpeningBracket(c); 

                        if (s.top() == opening_bracket) {
                                s.pop(); 
                        }
                        else {
                                return false; 
                        }
                }
                
        }

        return s.empty();
}
